File: evaluate-reverse-polish-notation/stack.py
import math
from typing import List


# leetcode.com boilerplate
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        return evaluate(tokens)


def evaluate(tokens):
    stack = []
    for token in tokens:
        if token == '+':
            b, a = stack.pop(), stack.pop()
            stack.append(a + b)
        elif token == '-':
            b, a = stack.pop(), stack.pop()
            stack.append(a - b)
        elif token == '*':
            b, a = stack.pop(), stack.pop()
            stack.append(a * b)
        elif token == '/':
            b, a = stack.pop(), stack.pop()
            stack.append(math.trunc(a / b))
        else:
            stack.append(int(token))

    return stack[0]

File: evaluate-reverse-polish-notation/stack.cpp
#include <cassert>
#include <cstdlib>
#include <string>
#include <vector>

int evaluate(const std::vector<std::string>& tokens) {
  std::vector<int> stack;

  for (const std::string& token : tokens) {
    const char first = token[0];
    if ((first >= '0' && first <= '9') || token.size() > 1) {
      stack.push_back(std::atoi(token.c_str()));
      continue;
    }

    // It's an operation.
    const int right = stack.back();
    stack.pop_back();
    const int left = stack.back();
    stack.pop_back();
    assert(token.size() == 1);
    int number;
    switch (first) {
    case '+':
      number = left + right;
      break;
    case '-':
      number = left - right;
      break;
    case '*':
      number = left * right;
      break;
    default:
      assert(first == '/');
      number = left / right;
      break;
    }
    stack.push_back(number);
  }

  return stack.back();
}

class Solution {
public:
    int evalRPN(std::vector<std::string>& tokens) {
      return evaluate(tokens);
    }
};
